Repair-PME.ps1

```diff
@@ -0,0 +1,105 @@
+                            If ($Install.ExitCode -eq 0) {
+                                Write-Output "$($PMEDetails.Name) version $($PMEDetails.Version) successfully installed"
+                            }
+                            Else {
+                                Write-Error "$($PMEDetails.Name) version $($PMEDetails.Version) was unable to be successfully installed, exit code $($Install.ExitCode)"
+                            }
+                    }
+                    Else {
+                        Write-Error "Hash of file downloaded ($($Download.Hash)) does not equal hash ($($PMEDetails.SHA256Checksum)) from sis.nable.com, aborting install"
+                    }
+            }
+    }
+    Else {
+        Write-Output "$($PMEDetails.FileName) does not exist, begin download and install phase"
+
+            # Check for PME Archive Directory
+            If (Test-Path "C:\ProgramData\SolarWinds MSP\PME\archives") {
+                Write-Output "C:\ProgramData\SolarWinds MSP\PME\archives already exists"
+            }
+            Else {
+                Write-Output "Directory 'C:\ProgramData\SolarWinds MSP\PME\archives' does not exist, creating directory"
+                New-Item -ItemType Directory -Path "C:\ProgramData\SolarWinds MSP\PME\archives" -Force | Out-Null
+            }
+
+        # Download Setup
+        Write-Output "Begin download of current $($PMEDetails.FileName) version $($PMEDetails.Version) from sis.a-able.com"
+        Start-BitsTransfer -Source "$($PMEDetails.DownloadURL)" -Destination "C:\ProgramData\SolarWinds MSP\PME\archives\$($PMEDetails.FileName)"
+        
+        # Check Hash
+        $Download = Get-LegacyHash -Path "C:\ProgramData\SolarWinds MSP\PME\archives\$($PMEDetails.FileName)" -Algorithm SHA256
+            If ($Download -eq $($PMEDetails.SHA256Checksum)) {
+                # Install
+                Write-Output "Hash of file is correct, installing $($PMEDetails.FileName)"
+                $Install = Start-process -filepath "C:\ProgramData\SolarWinds MSP\PME\archives\$($PMEDetails.FileName)" -argumentlist '/SP- /VERYSILENT /SUPPRESSMSGBOXES /NORESTART /CLOSEAPPLICATIONS' -wait -passthru
+                    If ($Install.ExitCode -eq 0) {
+                        Write-Output "$($PMEDetails.Name) version $($PMEDetails.Version) successfully installed"
+                    }
+                    Else {
+                        Write-Error "$($PMEDetails.Name) version $($PMEDetails.Version) was unable to be successfully installed, exit code $($Install.ExitCode)"
+                    }
+            }
+            Else {
+                Write-Error "Hash of file downloaded ($($Download.Hash)) does not equal hash ($($PMEDetails.SHA256Checksum)) from sis.nable.com, aborting install"
+            }
+    }
+}
+
+. Get-PMESetup
+. Install-PMESetup
```

Comment:
There could be one additional step, but it's a tricky one. Sometimes the issue with PME could be caused by some incorrect data in its data folders (C:\ProgramData\SolarWinds MSP\*.*). I would definitely recommend deleting all files  at least in following folders:
- C:\ProgramData\SolarWinds MSP\SolarWinds.MSP.CacheService\cache
- C:\ProgramData\SolarWinds MSP\SolarWinds.MSP.CacheService
I would rather keep all other files as they are, otherwise we could delete some user settings.

Reply:
Thanks - A new function has been created to perform this action.